JSON output for show-class-hierarchy

peterschrammel opened the pull request.

peterschrammel committed:
Add and unify --show-class-hierarchy command line option

thk123 commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -645,6 +645,16 @@ int jbmc_parse_optionst::get_goto_program(
+      hierarchy(lazy_goto_model.symbol_table);
+      show_class_hierarchy(
+        hierarchy, get_message_handler(), ui_message_handler.get_ui());
+      return 0;
```
`return CPROVER_EXIT_SUCCESS`

thk123 commented on src/goto-programs/class_hierarchy.cpp:
```diff
@@ -123,17 +126,17 @@ void class_hierarchyt::get_parents_trans_rec(
 }
 
-void class_hierarchyt::output(std::ostream &out) const
+void class_hierarchyt::output(std::ostream &out, bool children_only) const
```
Please document - it isn't obvious to me what `children_only` means without reading the code.

chrisr-diffblue replied:
General comment: (no need for any changes) - I know it's not really the coding style we have in the rest of the codebase, but personally I'd also like to see us use something more descriptive for these kinds of things as well, instead of a lot of `true` or `false` literals sprinkled around - but thats just me :-)

peterschrammel committed:
Improve class hierarchy output

more readable output, adds info whether class is abstract
removes the plain text unit test which is replaced by a
JBMC regression test in a later commit

peterschrammel committed:
Add JSON output for class hierarchy

martin-cs commented on src/goto-programs/class_hierarchy.cpp:
```diff
@@ -30,6 +31,9 @@ void class_hierarchyt::operator()(const symbol_tablet &symbol_table)
 
+      class_map[symbol_pair.first].is_abstract =
+        struct_type.get_bool(ID_abstract);
+
```
Feels like the kind of thing that could do with a test case.

peterschrammel replied:
Not sure what you want me to test here. Actually, that output has already proved useful to check whether we load interface/abstract class information correctly, but that can't really be put into a unit or regression test.

martin-cs replied:
Just a test that uses that part.  I ask because the code was right / finished before and now this needs to be added so I think we should have a case that was sub-par / wrong / insufficient before and is now better.  If it's a pain don't worry.

martin-cs reviewed (approved):
#seemslegit

chrisr-diffblue commented on src/goto-programs/class_hierarchy.cpp:
```diff
@@ -165,3 +166,29 @@ void class_hierarchyt::output_dot(std::ostream &ostr) const
+/// Output the class hierarchy in JSON format
+/// \param json_stream: the output JSON stream array
+/// \param children_only: print the children only and do not print the parents
+void class_hierarchyt::output(
```
Nit: It would be nice if the doxygen comment included an example of the expected JSON output/schema...

chrisr-diffblue replied:
Since this commit has added new functionality (the JSON output) - shouldn't this commit also include some unit/regression tests to exercise that output and validate that we are generating the expected JSON.

peterschrammel replied:
I've added tests for plain text and json in jbmc/regression/jbmc/class_hierarchy

chrisr-diffblue reviewed (approved):
A few minor comments/nits, but nothing blocking.
